### PR TITLE
feat(pki): Add support for ACME configuration

### PR DESCRIPTION
**Pull request by Viper61**
### Description

This PR adds a new resource to enable ACME configuration of a PKI (`vault_pki_secret_backend_config_acme`).

Relates #1947

### Checklist
- [x] Added [CHANGELOG](https://github.com/hashicorp/terraform-provider-vault/blob/master/CHANGELOG.md) entry (only for user-facing changes)
- [x] Acceptance tests where run against all supported Vault Versions

### Output from acceptance testing:

<details>
<summary>Vault 1.13</summary>

```
$ VAULT_ADDR=http://172.20.0.1:8213 TESTARGS="--run TestPkiSecretBackendConfigACME_basic" make testacc
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test --run TestPkiSecretBackendConfigACME_basic -timeout 30m ./...
?       github.com/hashicorp/terraform-provider-vault   [no test files]
?       github.com/hashicorp/terraform-provider-vault/cmd/coverage      [no test files]
?       github.com/hashicorp/terraform-provider-vault/cmd/generate      [no test files]
?       github.com/hashicorp/terraform-provider-vault/helper    [no test files]
?       github.com/hashicorp/terraform-provider-vault/internal/consts   [no test files]
ok      github.com/hashicorp/terraform-provider-vault/codegen   0.010s [no tests to run]
?       github.com/hashicorp/terraform-provider-vault/internal/identity/group   [no test files]
?       github.com/hashicorp/terraform-provider-vault/internal/identity/mfa     [no test files]
?       github.com/hashicorp/terraform-provider-vault/internal/pki      [no test files]
ok      github.com/hashicorp/terraform-provider-vault/internal/identity/entity  0.032s [no tests to run]
?       github.com/hashicorp/terraform-provider-vault/internal/sync     [no test files]
?       github.com/hashicorp/terraform-provider-vault/schema    [no test files]
ok      github.com/hashicorp/terraform-provider-vault/internal/provider 0.040s [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/testutil  0.006s [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/util      0.007s [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/vault     0.026s
```
</details>

<details>
<summary>Vault 1.14.0</summary>

```
$ VAULT_ADDR=http://172.20.0.1:8214 TESTARGS="--run TestPkiSecretBackendConfigACME_basic" make testacc
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test --run TestPkiSecretBackendConfigACME_basic -timeout 30m ./...
?       github.com/hashicorp/terraform-provider-vault   [no test files]
?       github.com/hashicorp/terraform-provider-vault/cmd/coverage      [no test files]
?       github.com/hashicorp/terraform-provider-vault/cmd/generate      [no test files]
?       github.com/hashicorp/terraform-provider-vault/helper    [no test files]
?       github.com/hashicorp/terraform-provider-vault/internal/consts   [no test files]
ok      github.com/hashicorp/terraform-provider-vault/codegen   0.011s [no tests to run]
?       github.com/hashicorp/terraform-provider-vault/internal/identity/group   [no test files]
?       github.com/hashicorp/terraform-provider-vault/internal/identity/mfa     [no test files]
?       github.com/hashicorp/terraform-provider-vault/internal/pki      [no test files]
ok      github.com/hashicorp/terraform-provider-vault/internal/identity/entity  0.025s [no tests to run]
?       github.com/hashicorp/terraform-provider-vault/internal/sync     [no test files]
?       github.com/hashicorp/terraform-provider-vault/schema    [no test files]
ok      github.com/hashicorp/terraform-provider-vault/internal/provider 0.028s [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/testutil  0.012s [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/util      0.005s [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/vault     3.727s
```
</details>

<details>
<summary>Vault 1.14</summary>

```
$ VAULT_ADDR=http://172.20.0.1:8314 TESTARGS="--run TestPkiSecretBackendConfigACME_basic" make testacc
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test --run TestPkiSecretBackendConfigACME_basic -timeout 30m ./...
?       github.com/hashicorp/terraform-provider-vault   [no test files]
?       github.com/hashicorp/terraform-provider-vault/cmd/coverage      [no test files]
?       github.com/hashicorp/terraform-provider-vault/cmd/generate      [no test files]
?       github.com/hashicorp/terraform-provider-vault/helper    [no test files]
?       github.com/hashicorp/terraform-provider-vault/internal/consts   [no test files]
ok      github.com/hashicorp/terraform-provider-vault/codegen   0.006s [no tests to run]
?       github.com/hashicorp/terraform-provider-vault/internal/identity/group   [no test files]
?       github.com/hashicorp/terraform-provider-vault/internal/identity/mfa     [no test files]
?       github.com/hashicorp/terraform-provider-vault/internal/pki      [no test files]
ok      github.com/hashicorp/terraform-provider-vault/internal/identity/entity  0.049s [no tests to run]
?       github.com/hashicorp/terraform-provider-vault/internal/sync     [no test files]
?       github.com/hashicorp/terraform-provider-vault/schema    [no test files]
ok      github.com/hashicorp/terraform-provider-vault/internal/provider 0.030s [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/testutil  0.006s [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/util      0.004s [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/vault     3.848s
```
</details>

<details>
<summary>Vault 1.15</summary>

```
$ VAULT_ADDR=http://172.20.0.1:8215 TESTARGS="--run TestPkiSecretBackendConfigACME_basic" make testacc
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test --run TestPkiSecretBackendConfigACME_basic -timeout 30m ./...
?       github.com/hashicorp/terraform-provider-vault   [no test files]
?       github.com/hashicorp/terraform-provider-vault/cmd/coverage      [no test files]
?       github.com/hashicorp/terraform-provider-vault/cmd/generate      [no test files]
?       github.com/hashicorp/terraform-provider-vault/helper    [no test files]
?       github.com/hashicorp/terraform-provider-vault/internal/consts   [no test files]
ok      github.com/hashicorp/terraform-provider-vault/codegen   0.007s [no tests to run]
?       github.com/hashicorp/terraform-provider-vault/internal/identity/group   [no test files]
?       github.com/hashicorp/terraform-provider-vault/internal/identity/mfa     [no test files]
?       github.com/hashicorp/terraform-provider-vault/internal/pki      [no test files]
?       github.com/hashicorp/terraform-provider-vault/internal/sync     [no test files]
?       github.com/hashicorp/terraform-provider-vault/schema    [no test files]
ok      github.com/hashicorp/terraform-provider-vault/internal/identity/entity  0.048s [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/internal/provider 0.055s [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/testutil  0.007s [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/util      0.005s [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/vault     3.807s
```
</details>

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

### CODE REVIEW

**Comment by igor-nikiforov** [comment]
@fairclothjm Could you please review this PR? Thanks!

**Comment by igor-nikiforov** [comment]
@benashz friendly reminder.

File: vault/resource_pki_secret_backend_config_acme.go [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+				ValidateFunc: validation.StringInSlice(
+					[]string{"not-required", "new-account-required", "always-required"}, false),
+			},
+			consts.FieldDnsResolver: {
+				Type:     schema.TypeString,
+				Optional: true,
+				Description: "DNS resolver to use for domain resolution on this mount. " +
+					"Must be in the format <host>:<port>, with both parts mandatory.",
+				ValidateFunc: validation.StringMatch(
+					regexp.MustCompile(`^([0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}:[0-9]{1,5})?$`), ""),
```
[the remark this answers is not in the document]
Comment (Viper61) [reply vault/resource_pki_secret_backend_config_acme.go]: True! The ValidateFunc is removed.

File: internal/consts/consts.go [hunk 2]
```diff
@@ -456,14 +463,15 @@ const (
...
-	VaultVersion114 = "1.14.0"
-	VaultVersion115 = "1.15.0"
-	VaultVersion116 = "1.16.0"
+	VaultVersion190  = "1.9.0"
+	VaultVersion110  = "1.10.0"
+	VaultVersion111  = "1.11.0"
+	VaultVersion112  = "1.12.0"
+	VaultVersion113  = "1.13.0"
+	VaultVersion114  = "1.14.0"
+	VaultVersion1141 = "1.14.1"
```
[the remark this answers is not in the document]
Comment (Viper61) [reply internal/consts/consts.go]: This was added become `allow_role_ext_key_usage` was introduced in 1.14.1.
There is a note on `pki_secret_backend_acme_bck/website/docs/r/pki_secret_backend_config_acme.html.md` to indicate this field is for 1.14.1+ only.
Is it enough?

File: vault/resource_pki_secret_backend_config_acme.go [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+	}
+)
+
+func pkiSecretBackendConfigACMEResource() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: provider.MountCreateContextWrapper(pkiSecretBackendConfigACMECreate, provider.VaultVersion113),
+		ReadContext:   provider.ReadContextWrapper(pkiSecretBackendConfigACMERead),
+		UpdateContext: pkiSecretBackendConfigACMEUpdate,
+		DeleteContext: pkiSecretBackendConfigACMEDelete,
+		Importer: &schema.ResourceImporter{
```
[the remark this answers is not in the document]
Comment (Viper61) [reply vault/resource_pki_secret_backend_config_acme.go:39]: Absolutely. Thank you.

File: vault/resource_pki_secret_backend_config_acme.go [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+				Type:        schema.TypeBool,
+				Required:    true,
+				Description: "Specifies whether ACME is enabled.",
+			},
+			consts.FieldDefaultDirectoryPolicy: {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Computed:    true,
+				Description: "Specifies the policy to be used for non-role-qualified ACME requests.",
+				ValidateFunc: validation.StringMatch(
```
[the remark this answers is not in the document]
Comment (Viper61) [reply vault/resource_pki_secret_backend_config_acme.go]: As requested, all ValidateFunc were removed.

File: vault/resource_pki_secret_backend_config_acme.go [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+	log.Printf("[DEBUG] Reading %s from Vault", path)
+	resp, err := client.Logical().ReadWithContext(ctx, path)
+	if err != nil {
+		return diag.Errorf("error reading from Vault: %s", err)
+	}
+
+	if resp == nil {
+		d.SetId("")
+
+		return nil
```
[the remark this answers is not in the document]
Comment (Viper61) [reply vault/resource_pki_secret_backend_config_acme.go]: This is updated with an error message.

**Comment by Viper61** [comment]
Hi @stevendpclark, Thanks for your review.
The rebase and requested updates are complete.
Let me know if you think any further changes are needed.

File: internal/consts/consts.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (stevendpclark) [reply internal/consts/consts.go]: Yeah I'd say so, we are beyond the support lifecycle anyways now for 1.14.

**Review by stevendpclark — APPROVED** [review]
Looks great to me! Thanks @Viper61 for making all the requested changes.

**Pull request merged**